Isolate recycler behavior in PV E2E

copejon opened the pull request:
**What this PR does / why we need it**:
Sets the default `reclaimPolicy` for PV E2E to `Retain` and isolates `Recycle` tests to their own context.  The purpose of this is to future proof the PV test suite against the possible deprecation of the `Recycle` behavior.  This is done by consolidating recycling test code into a single Context block that can be removed en masse without affecting the test suite.

Secondly, adds a liveliness check for the NFS server pod prior to each test to avoid maxing out timeouts if the NFS server becomes unavailable.

cc @saad-ali @jeffvance

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40177)
<!-- Reviewable:end -->

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -49,19 +50,21 @@ type pvmap map[string]pvval
+//   reclaimPolicy (Optional) makePersistentVolume() defaults to Retain if unset
+// pvSource: v1.PersistentVolumeSource{
 // 	NFS: &api.NFSVolumeSource{
 // 		.
```
why the blank lines here and below?

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -119,10 +122,10 @@ func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns strin
+func deletePVCandValidatePV(c clientset.Interface, ns string, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume, expectPVPhase v1.PersistentVolumePhase) {
 
 	pvname := pvc.Spec.VolumeName
 	framework.Logf("Deleting PVC %v to trigger recycling of PV %v", pvc.Name, pvname)
```
recycling only triggered if the pv's reclaim policy is "Recycle". So for the case of "Retain" this message is misleading.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -457,7 +450,7 @@ func completeTest(f *framework.Framework, c clientset.Interface, ns string, pv *
 // Note: this func is serialized, we wait for each pod to be deleted before creating the
 //   next pod. Adding concurrency is a TODO item.
 // Note: this func is called recursively when there are more claims than pvs.
```
AFAICT this func is no longer recursive. That may have changed a while ago...

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -522,6 +515,14 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
 				serverIP = nfsServerPod.Status.PodIP
 				framework.Logf("NFS server IP address: %v", serverIP)
+			} else {
+				// If the nfs server does exist, check that it is Running and the volume is exported.
```
These are good tests. See also pr [38214](https://github.com/kubernetes/kubernetes/pull/38214/files#diff-c83ebba11ffd12bd594ac8aa2f9301ff) where it is no longer an error if the server pod already exists.

copejon replied:
Per our Fri conversation:  I'm going to open another PR to refactor the pattern we use for creating / destroying the NFS server pod.  We will instead create a new pod for each test spec within the test's namespace.  This will increase the robustness of each test since they will have a dedicated volume to test against.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -617,9 +618,10 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
 			// Note: PVs are created before claims and no pre-binding
 			It("should create 2 PVs and 4 PVCs: test write access [Volume][Serial][Flaky]", func() {
 				numPVs, numPVCs := 2, 4
+				pvConfig.reclaimPolicy = v1.PersistentVolumeReclaimRecycle
```
So "recycle" is not fully isolated.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -446,9 +439,9 @@ func completeTest(f *framework.Framework, c clientset.Interface, ns string, pv *
 	createWaitAndDeletePod(f, c, ns, pvc.Name)
 
-	// 3. delete the PVC, wait for PV to become "Available"
+	// 3. delete the PVC, wait for PV to become "Released"
```
Line 327 has a dangling comment, just fyi.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +639,55 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+			// This It() tests a scenario where a PV is written to by a Pod, recycled, then the volume checked
+			// for files. If files are found, the checking Pod fails, failing the test.  Otherwise, the pod
+			// (and test) succeed.
+			It("should test that a PV becomes Available and is clean after the PVC is deleted. [Jon][Volume][Serial][Flaky]", func() {
```
[Jon]  :)

copejon replied:
whoops!

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +639,55 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+				pod := makeWritePod(ns, pvc.Name)
+				pod, err := c.Core().Pods(ns).Create(pod)
+				Expect(err).NotTo(HaveOccurred())
+				framework.WaitForPodSuccessInNamespace(c, pod.Name, ns)
```
no assignment to and check for err?

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +639,55 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+
+				// If a file is detected in /mnt, fail the pod and do not restart it.
+				By("Verifying the mount has been cleaned.")
+				mount := pod.Spec.Containers[0].VolumeMounts[0].MountPath
```
maybe we should have a common way/pattern to test if a volume's mount is present on the target pod's host? I do something similar in pr  [38214](https://github.com/kubernetes/kubernetes/pull/38214/files#diff-e998808179b7c27b4062f9fa3b7bfa3eR253)

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +639,55 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+				By("Verifying the mount has been cleaned.")
+				mount := pod.Spec.Containers[0].VolumeMounts[0].MountPath
+				pod = makePod(ns, pvc.Name, fmt.Sprintf("[ $(ls -A %s | wc -l) -eq 0 ] && exit 0 || exit 1", mount))
+				pod.Spec.RestartPolicy = v1.RestartPolicyNever
```
The restart policy in the template is `v1.RestartPolicyOnFailure`. Do we want that or is it better to set the restart policy to `v1.RestartPolicyNever` for all pods in the test?

copejon replied:
Agreed, we Never should be the default policy for pods.  Well alter the template instead.

ixdy commented:
@k8s-bot ok to test

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -119,48 +117,47 @@ func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns strin
 
-// Delete the PVC and wait for the PV to become Available again. Validate that the PV
+// Delete the PVC and wait for the PV to enter its expected phase. Validate that the PV
 // has recycled (assumption here about reclaimPolicy). Caller tells this func which
```
I think the sentence about validating that the pv has recycled is a bit misleading since in most cases the pv will not be recycled. I think it's ok to just say _and wait for the PV to enter its expected phase_.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +623,56 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+				By("Re-mounting the volume.")
+				pvc = makePersistentVolumeClaim(ns)
+				pvc = createPVC(c, ns, pvc)
+				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, c, ns, pvc.Name, 2*time.Second, 60*time.Second)
```
Need to handle err != nil here.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -637,7 +623,56 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+			// (and test) succeed.
+			It("should test that a PV becomes Available and is clean after the PVC is deleted. [Volume][Serial][Flaky]", func() {
+				By("Writing to the volume.")
+				pod := makeWritePod(ns, pvc.Name)
```
L653-657: can you call `createClientPod` here?

copejon replied:
no, but we can write a couple new functions to mirror the pattern of makeWritePod wrapping makePod.  Alternatively, we could use createClientPod then podExec to make a file on the volume

spxtr commented:
/approve

smarterclayton commented:
/approve

copejon commented:
@jsafrane  can of you PTAL for an lgtm?

jeffvance commented:
@copejon my pr #40530 does significant restructuring to _persistent_volumes.go_. Let me know if you want any help with the merge.

jsafrane commented:
```
I0201 16:03:09.511] Verifying hack/make-rules/../../hack/verify-gofmt.sh
I0201 16:03:23.545] !!! 'gofmt -s -w' needs to be run on the following files: 
I0201 16:03:23.545] ./test/e2e/persistent_volumes.go
```

jsafrane commented on test/e2e/pvutil.go:
```diff
@@ -130,35 +131,35 @@ func deletePVCandValidatePV(c clientset.Interface, ns string, pvc *v1.Persistent
 	Expect(err).NotTo(HaveOccurred())
 
-	// examine the pv's ClaimRef and UID and compare to expected values
+	// examine the pv's ClaimRef and UID and compa re to expected values
```
extra space `compa re`?

jsafrane reviewed (commented):
one minor nit and failed verification - fix those and it's good to go.

copejon commented:
@jsafrane thanks.  fixed typo + rebased.

jsafrane commented:
/lgtm

jsafrane commented:
@jeffvance, it seems it's up to you to mark it as lgtml. We'll see if you have enough right.

deads2k commented:
/approve

jeffvance commented:
Better isolation is acheived by creating a new  _pv_recycle.go_ file which can be deleted once recycling is deprecated. This is consistent with new files created by #40756.  Thoughts?

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -41,18 +42,17 @@ func completeTest(f *framework.Framework, c clientset.Interface, ns string, pv *
 	createWaitAndDeletePod(f, c, ns, pvc.Name)
 
-	// 3. delete the PVC, wait for PV to become "Available"
+	// 3. delete the PVC, wait for PV to become "Released"
```
L46, L73: _By("Deleting the PVC to invoke the recycler")_. `completeTest` and `completeMultiTest` do not invoke recycling since the pv's reclaimPolicy is set to "Retain" for these test cases.

jeffvance replied:
`completeTest` and `completeMultiTest` still both mention "triggering the recycler" even though that only occurs in a single Context in _persistent_volumes.go_

jeffvance commented on test/e2e/pvutil.go:
```diff
@@ -117,10 +118,10 @@ func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns strin
 
-// Delete the PVC and wait for the PV to become Available again. Validate that the PV
+// Delete the PVC and wait for the PV to enter its expected phase. Validate that the PV
 // has recycled (assumption here about reclaimPolicy). Caller tells this func which
```
I'd delete the "recycle" comment. Simply validating that the pv attains its expected Phase.

jeffvance replied:
"recycle" still appears in the comment, even though this is a minor case now.

jeffvance commented on test/e2e/pvutil.go:
```diff
@@ -130,35 +131,35 @@ func deletePVCandValidatePV(c clientset.Interface, ns string, pvc *v1.Persistent
 
-	// Wait for the PV's phase to return to Available
+	// Wait for the PV's phase to return to be `expectPVPhase`
 	framework.Logf("Waiting for recycling process to complete.")
```
reword this to remove "recycling".

jeffvance commented on test/e2e/pvutil.go:
```diff
@@ -130,35 +131,35 @@ func deletePVCandValidatePV(c clientset.Interface, ns string, pvc *v1.Persistent
+	} else if expectPVPhase == v1.VolumeBound {
 		Expect(cr).NotTo(BeNil())
 		Expect(len(cr.UID)).NotTo(BeZero())
 	}
```
do we need an `else if` for "Released"?

copejon replied:
when released, no claimRef fields are altered so the check wouldn't find anything interesting.

jeffvance commented:
The AWS e2e test has been rerun several times. It panics:
```
...
STEP: Verifying that the 2nd pod wont be removed if it is not running and ready
I0203 15:47:38.773] Feb  3 15:45:52.719: INFO: Verifying statefulset ss doesn't scale past 2 for another 9.99999975s
I0203 15:47:38.773] Feb  3 15:45:53.771: INFO: Verifying statefulset ss doesn't scale past 2 for another 8.944747298s
I0203 15:47:38.773] Feb  3 15:45:54.824: INFO: Verifying statefulset ss doesn't scale past 2 for another 7.891875811s
I0203 15:47:38.773] Feb  3 15:45:55.877: INFO: Verifying statefulset ss doesn't scale past 2 for another 6.839165692s
I0203 15:47:38.773] Feb  3 15:45:56.929: INFO: Verifying statefulset ss doesn't scale past 2 for another 5.786701972s
I0203 15:47:38.773] Feb  3 15:45:58.019: INFO: Verifying statefulset ss doesn't scale past 2 for another 4.734414321s
I0203 15:47:38.774] Feb  3 15:45:59.096: INFO: Verifying statefulset ss doesn't scale past 2 for another 3.644001338s
I0203 15:47:38.774] Feb  3 15:46:00.173: INFO: Verifying statefulset ss doesn't scale past 2 for another 2.567293372s
I0203 15:47:38.774] Feb  3 15:46:01.227: INFO: Verifying statefulset ss doesn't scale past 2 for another 1.489983667s
I0203 15:47:38.774] Feb  3 15:46:02.281: INFO: Verifying statefulset ss doesn't scale past 2 for another 436.139835ms
I0203 15:47:38.774] STEP: Verifying the 2nd pod is removed only when it becomes running and ready
I0203 15:47:38.774] Feb  3 15:46:03.448: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.aws-kops-agent-pr-79-0.test-aws.k8s.io --kubeconfig=/tmp/kops-kubecfg059554481 exec --namespace=e2e-tests-statefulset-rbr31 ss-0 -- /bin/sh -c mv -v /tmp/index.html /usr/share/nginx/html/'
I0203 15:47:38.774] Feb  3 15:46:04.317: INFO: stderr: ""
I0203 15:47:38.774] Feb  3 15:46:04.318: INFO: stdout of mv -v /tmp/index.html /usr/share/nginx/html/ on ss-0: '/tmp/index.html' -> '/usr/share/nginx/html/index.html'
I0203 15:47:38.774] 
I0203 15:47:38.775] Feb  3 15:46:04.318: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.aws-kops-agent-pr-79-0.test-aws.k8s.io --kubeconfig=/tmp/kops-kubecfg059554481 exec --namespace=e2e-tests-statefulset-rbr31 ss-1 -- /bin/sh -c mv -v /tmp/index.html /usr/share/nginx/html/'
I0203 15:47:38.775] Feb  3 15:46:05.514: INFO: stderr: ""
I0203 15:47:38.775] Feb  3 15:46:05.515: INFO: stdout of mv -v /tmp/index.html /usr/share/nginx/html/ on ss-1: '/tmp/index.html' -> '/usr/share/nginx/html/index.html'
I0203 15:47:38.775] 
I0203 15:47:38.775] [AfterEach] [k8s.io] Basic StatefulSet functionality
I0203 15:47:38.775]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/statefulset.go:107
...
•! Panic [210.031 seconds]
I0203 15:47:38.823] [k8s.io] StatefulSet
I0203 15:47:38.823] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:656
I0203 15:47:38.824]   [k8s.io] Basic StatefulSet functionality
I0203 15:47:38.824]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:656
I0203 15:47:38.824]     Scaling down before scale up is finished should wait until current pod will be running and ready before it will be removed [It]
I0203 15:47:38.824]     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/statefulset.go:274
I0203 15:47:38.824] 
I0203 15:47:38.824]     Test Panicked
I0203 15:47:38.824]     interface conversion: runtime.Object is *v1.Status, not *v1.Pod
I0203 15:47:38.824]     /usr/local/go/src/runtime/panic.go:458
I0203 15:47:38.824] 
I0203 15:47:38.825]     Full Stack Trace
I0203 15:47:38.825]     	/usr/local/go/src/runtime/panic.go:458 +0x243
I0203 15:47:38.825]     k8s.io/kubernetes/test/e2e.glob..func82.2.6.1(0xc420f311a0, 0x5, 0x3a626a0, 0xc4209ad000, 0x1, 0x28d556a, 0x28d5569)
I0203 15:47:38.825]     	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/statefulset.go:259 +0x48b
I0203 15:47:38.825]     k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch.Until(0x8bb2c97000, 0x3a6dd60, 0xc4202e1b90, 0xc420b1f2a0, 0x1, 0x1, 0x0, 0x0, 0x0)
I0203 15:47:38.825]     	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch/until.go:69 +0x25b
I0203 15:47:38.825]     k8s.io/kubernetes/test/e2e.glob..func82.2.6()
I0203 15:47:38.825]     	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/statefulset.go:272 +0xbf3
I0203 15:47:38.825]     k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc420f846c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
I0203 15:47:38.826]     	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:257 +0x232
I0203 15:47:38.826]     k8s.io/kubernetes/test/e2e.TestE2E(0xc420720600)
I0203 15:47:38.826]     	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:30 +0x2b
I0203 15:47:38.826]     testing.tRunner(0xc420720600, 0x28dd2e0)
I0203 15:47:38.826]     	/usr/local/go/src/testing/testing.go:610 +0x81
I0203 15:47:38.826]     created by testing.(*T).Run
I0203 15:47:38.826]     	/usr/local/go/src/testing/testing.go:646 +0x2ec
```

jeffvance commented:
@k8s-bot kops aws e2e test this

copejon commented:
@jeffvance PTAL

jeffvance commented on test/e2e/pvutil.go:
```diff
@@ -117,10 +118,10 @@ func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns strin
+func deletePVCandValidatePV(c clientset.Interface, ns string, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume, expectPVPhase v1.PersistentVolumePhase) {
 
 	pvname := pvc.Spec.VolumeName
 	framework.Logf("Deleting PVC %v to trigger recycling of PV %v", pvc.Name, pvname)
```
"recycling" is not triggered unless reclaimPolicy is "Recycle", and that is limited to a single Context now.

copejon committed:
PV E2E: default reclaimPolicy to Retain, target recycling behavior in single spec

Added nfs-server verification between tests; prototyped recycler test

recycler test in working state, needs clean up/comments

comments added; gets mount path from pod instead of hard code

typos

removed NFS server checker, will be handled in separate PR

catch err; inaccurate comment

gofmt; typo

jsafrane commented:
/lgtm
